### DISCUSSION TITLE
NCA analysis using PKNCA - errors and different results than expected

### DISCUSSION THREAD

**Discussion by caribap** [humanpred/pknca#283] [Q&A] [answered]
Rversion = 4.3.2
I want to perform an NCA analysis for a single dose studies. I have concentration measurements at 5 timepoints for each subject, and have 3 subjects for each single dose - 1) 1mg/kg IV, 2) 9mg/kg IV and 3) 15mg/kg PO. In the end I want to have a table with pk parameters for the mean between the 3 subjects at each single dose study. 
I am encountering 3 problems: 
- when I execute iv_conc <- PKNCAconc(iv_data, Concentration~Time|Dose) I get an Error: Error in PKNCAconc.data.frame(as.data.frame(data), ...) : 
  Rows that are not unique per group and time (column names: Time, Dose) found within concentration data.  Row numbers: 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48
- I get my results for each subject while I want the mean for each study group
- I get different results from the phoenix NCA analysis (significantly different)

Note: I am using asnumeric and is.na to make sure my dose, concentration and time are manipulated correctly; I am using read_excel() to read my data and in the beginning I split my data by route creating 2 data frames 

Thank you so much for your help!

**Comment by billdenney** [comment, marked as answer]
The first issue is the main one.  Without fixing it, I'm not sure how you got results of any type.

The way that you specified the call to PKNCAconc, you didn't indicate a subject column.  Given your description of the study, that is the most likely issue.  The way to specify the subject column is to add the subject as the last grouping level similar to the.below call:

```
iv_conc <- PKNCAconc(iv_data, Concentration~Time|Dose+Subject)
```

To get the group mean, you would run `summary()` on your NCA results object.

For the difference with Phoenix, I would need to know more about what parameters you are calculating and your data.  Prior tests comparing PKNCA to Phoenix have found almost no differences, and where the differences exist, they are intentional (AUC for all 0 concentrations and an edge case of half lives around order of automatically selecting the best half life).

**Reply by caribap** [reply]
Thank you! But the reason I did leave Subject aside was because by running the line of code you provided I get another error: Error in (function (..., row.names = NULL, check.rows = FALSE, check.names = TRUE,  : 
  arguments imply differing number of rows: 48, 0, 1
By running iv_conc <- PKNCAconc(iv_data, Concentration~Time|Subject+Dose) though, this error do not appear. But, I do not get results because when I run: 

iv_data <- PKNCAdata(iv_conc, iv_dose, intervals=d_interval)
o_result <- pk.nca(iv_data)
o_result 

I get this error: Error in standardize_column_names(x = data_no_exclude, cols = cols, group_cols = group_cols,  : 
  group_cols must not overlap with other column names.  Change the name of the following groups: Dose

Plus when I do summary I get a mean of my 5 time points for each mice, and that is not what I want. I do not have a group column that specifies that mice 1-3 (1mg/kg IV) is group 1, should I do that?
Thank you for your quick response and value help!

**Reply by billdenney** [reply]
It seems like there is more going on here with interactions between different parts of PKNCA.  I think that there is an issue with the way that you are specifying the grouping.

Can you please share your full code?  If possible, can you please share the data, too?

**Reply by caribap** [reply]
I am not able to share my full data but I can give you a more detailed description: so I have the following column names:  Study, Species, Strain, Sex, Compound, Formulation, Subject, Dose, Route, Time, Concentration. Lets say mice 1 received a dose of 1 mg/kg IV - I will collect samples at time points 1h, 2h, 4h, 8h and 24h and for each I have a concentration of lets say 1000ng/mL, 400ng/mL, 200ng/mL, 80ng/mL and BLOQ. then for mice 2 and 3 it's the same reasoning. For mice 4, 5 and 6 they received a dose of 9 mg/kg IV and for mice 7,8 and 9 they  received a dose of 15 mg/kg PO. 

That is how I wrote my code:
library.List <- c("plyr","dplyr","tidyr","readxl","ggplot2", "PKNCA", "purrr")
for (pack in library.List){
  if(!require(pack, character.only=TRUE)){
    install.packages(pack)
    library(pack, character.only=TRUE)
  }
}

# Set the working directory
setwd("...")

# Define the R script file path
rscript_path <- file.path(getwd(), "PK_analysis.R")

# Specify the output file path
output_path <- file.path(getwd(), "PK_analysis.png")

# Load dataset
pk_data <- read_excel("...", skip = 0)

# Visualise the first five rows of my data
head(pk_data)

#---------- Data Preparation ---------------------------------------------------------------------------
# Assuming 'IV' for intravascular and 'PO' for extravascular routes
iv_data <- pk_data %>% filter(Route == "IV")
oral_data <- pk_data %>% filter(Route == "PO")

# Convert values to numeric for both datasets
iv_data$Concentration <- as.numeric(iv_data$Concentration)
iv_data$Dose <- as.numeric(iv_data$Dose)
iv_data$Time <- as.numeric(iv_data$Time)

oral_data$Concentration <- as.numeric(oral_data$Concentration)
oral_data$Dose <- as.numeric(oral_data$Dose)
oral_data$Time <- as.numeric(oral_data$Time)

# Replace NA values resulting from BLQ with zero
iv_data$Concentration[is.na(iv_data$Concentration)] <- 0
oral_data$Concentration[is.na(oral_data$Concentration)] <- 0

--------
# Define a concentration-time object
iv_conc <- PKNCAconc(iv_data, Concentration~Time|Subject+Dose, sparse=TRUE)
head(iv_conc)

# Define a dose object
iv_dose <- PKNCAdose(iv_data, Dose~Time|Subject, "intravascular", sparse=TRUE)
iv_dose

# Define intervals
d_interval <- data.frame(start = 0.083,
                         end = 24,
                         c0 = TRUE,
                         cmax = TRUE,
                         tmax = TRUE,
                         Tlast = TRUE,
                         aucivlast = TRUE,
                         aucinf.obs = TRUE
                         # clr.obs = TRUE,
                         # half.life = TRUE,
                         # vd.obs = TRUE
                         )

# Merge concentration and dose objects with intervals
iv_data <- PKNCAdata(iv_conc,iv_dose, intervals=d_interval)

# Calculate the NCA results
o_result <- pk.nca(iv_data)
o_result
 
# Display the results
pander::pander(summary(o_result))
pander::pander(head(as.data.frame(o_result), n = 126))

(and then I want to do the same for PO.)

Thank you for your help!

**Reply by caribap** [reply]
I have been playing around with grouping but I am now I get this 2 recurrent errors when I run o_result <- pk.nca(iv_data) :
- Error in standardize_column_names(x = data_no_exclude, cols = cols, group_cols = group_cols,  : 
  group_cols must not overlap with other column names.  Change the name of the following groups: Dose
- Error in `purrr::pmap()`:
ℹ In index: 1.
Caused by error in `vapply()`:
! values must be length 1,
 but FUN(X[[6]]) result is length 0
Run `rlang::last_trace()` to see where the error occurred.

And I don't get results anymore.

Thank you for your help!

**Reply by billdenney** [reply]
The issue is due to the fact that you are using `Dose` as a grouping variable and to indicate the dose.  (The names are the same which causes some internal data conflicts within PKNCA.  I'll try to fix this in a future release.)

So, if you could create a new variable to use for your groups other than `Dose`, that should fix it:

```
iv_data <-
  iv_data |>
  mutate(DoseGroup = Dose)
iv_conc <- PKNCAconc(iv_data, Concentration~Time|Subject+DoseGroup, sparse=TRUE)
```

**Reply by caribap** [reply]
Thank you! That helps, but I still get the Error in purrr::pmap():
ℹ In index: 1.
Caused by error in vapply():
! values must be length 1,
but FUN(X[[6]]) result is length 0
Run rlang::last_trace() to see where the error occurred.
which is forbidding me to get results, do you have any tips for that?

Thanks!

**Reply by billdenney** [reply]
That error is not clear to me where it's coming from.

Can you try anonymizing your data and sending it?  One way to anonymize the data is by doing something like randomizing your concentration data with something like this:

```
iv_data$Concentration <- iv_data$Concentration*exp(rnorm(nrow(iv_data)))
```

**Reply by caribap** [reply]
I have prepared a new file with made up values but that respects the structure I am using. I am not able to upload the file here for some reason - so I will leave here  the results.

Study | Species | Strain | Sex | Compound | Subject | Group | Dose | Analyte | DoseTime | Route | Tissue | Time | Concentration
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
  |   |   |   |   |   |   | mg/kg |   |   |   |   | h | ng/mL
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 62 | 1 | 1 |   | 0 | IV |   | 0.083 | 1070
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 62 | 1 | 1 |   |   | IV |   | 0.25 | 400
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 62 | 1 | 1 |   |   | IV |   | 0.50 | 148
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 62 | 1 | 1 |   |   | IV |   | 1 | 8.16
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 62 | 1 | 1 |   |   | IV |   | 2 | BLOQ
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 62 | 1 | 1 |   |   | IV |   | 4 | BLOQ
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 62 | 1 | 1 |   |   | IV |   | 8 | BLOQ
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 62 | 1 | 1 |   |   | IV |   | 24 | BLOQ
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 63 | 1 | 1 |   | 0 | IV |   | 0.083 | 1290
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 63 | 1 | 1 |   |   | IV |   | 0.25 | 501
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 63 | 1 | 1 |   |   | IV |   | 0.50 | 212
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 63 | 1 | 1 |   |   | IV |   | 1 | 13.9
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 63 | 1 | 1 |   |   | IV |   | 2 | BLOQ
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 63 | 1 | 1 |   |   | IV |   | 4 | BLOQ
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 63 | 1 | 1 |   |   | IV |   | 8 | BLOQ
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 63 | 1 | 1 |   |   | IV |   | 24 | BLOQ
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 64 | 1 | 1 |   | 0 | IV |   | 0.083 | 1008
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 64 | 1 | 1 |   |   | IV |   | 0.25 | 250
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 64 | 1 | 1 |   |   | IV |   | 0.50 | 130
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 64 | 1 | 1 |   |   | IV |   | 1 | 6.00
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 64 | 1 | 1 |   |   | IV |   | 2 | BLOQ
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 64 | 1 | 1 |   |   | IV |   | 4 | BLOQ
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 64 | 1 | 1 |   |   | IV |   | 8 | BLOQ
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 64 | 1 | 1 |   |   | IV |   | 24 | BLOQ
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 65 | 2 | 10 |   | 0 | IV |   | 0.083 | 7740
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 65 | 2 | 10 |   |   | IV |   | 0.25 | 2980
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 65 | 2 | 10 |   |   | IV |   | 0.50 | 1119
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 65 | 2 | 10 |   |   | IV |   | 1 | 140
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 65 | 2 | 10 |   |   | IV |   | 2 | 26.9
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 65 | 2 | 10 |   |   | IV |   | 4 | 11.4
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 65 | 2 | 10 |   |   | IV |   | 8 | BLOQ
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 65 | 2 | 10 |   |   | IV |   | 24 | BLOQ
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 66 | 2 | 10 |   | 0 | IV |   | 0.083 | 8600
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 66 | 2 | 10 |   |   | IV |   | 0.25 | 1700
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 66 | 2 | 10 |   |   | IV |   | 0.50 | 1800
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 66 | 2 | 10 |   |   | IV |   | 1 | 248
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 66 | 2 | 10 |   |   | IV |   | 2 | 44.8
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 66 | 2 | 10 |   |   | IV |   | 4 | 3.67
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 66 | 2 | 10 |   |   | IV |   | 8 | 3.77
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 66 | 2 | 10 |   |   | IV |   | 24 | BLOQ
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 67 | 2 | 10 |   | 0 | IV |   | 0.083 | 9200
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 67 | 2 | 10 |   |   | IV |   | 0.25 | 4640
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 67 | 2 | 10 |   |   | IV |   | 0.50 | 1720
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 67 | 2 | 10 |   |   | IV |   | 1 | 240
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 67 | 2 | 10 |   |   | IV |   | 2 | 20.3
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 67 | 2 | 10 |   |   | IV |   | 4 | 1.08
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 67 | 2 | 10 |   |   | IV |   | 8 | 5.87
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 67 | 2 | 10 |   |   | IV |   | 24 | BLOQ
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 68 | 3 | 10 |   | 0 | PO |   | 0.25 | 3.60
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 68 | 3 | 10 |   |   | PO |   | 0.50 | 3.59
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 68 | 3 | 10 |   |   | PO |   | 1 | 4.02
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 68 | 3 | 10 |   |   | PO |   | 2 | 3.00
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 68 | 3 | 10 |   |   | PO |   | 4 | BLOQ
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 68 | 3 | 10 |   |   | PO |   | 8 | BLOQ
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 68 | 3 | 10 |   |   | PO |   | 24 | BLOQ
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 69 | 3 | 10 |   | 0 | PO |   | 0.25 | 7.47
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 69 | 3 | 10 |   |   | PO |   | 0.50 | 4.01
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 69 | 3 | 10 |   |   | PO |   | 1 | 4.01
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 69 | 3 | 10 |   |   | PO |   | 2 | 1.72
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 69 | 3 | 10 |   |   | PO |   | 4 | BLOQ
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 69 | 3 | 10 |   |   | PO |   | 8 | BLOQ
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 69 | 3 | 10 |   |   | PO |   | 24 | BLOQ
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 70 | 3 | 10 |   | 0 | PO |   | 0.25 | 5.71
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 70 | 3 | 10 |   |   | PO |   | 0.50 | 5.77
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 70 | 3 | 10 |   |   | PO |   | 1 | 8.46
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 70 | 3 | 10 |   |   | PO |   | 2 | 3.88
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 70 | 3 | 10 |   |   | PO |   | 4 | 1.48
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 70 | 3 | 10 |   |   | PO |   | 8 | BLOQ
PK study x | Mouse | ICR(CD-1) | Male | Compound 36978 | Mouse 70 | 3 | 10 |   |   | PO |   | 24 | BLOQ

I performed a NCA analysis with phoenix and these are the results I am expecting:

NCA IV:
Rsq_adjusted | HL_Lambda_z (h) | Cmax | Cmax_D (kg*ng/mL/mg) | C0 (ng/mL) | Tlast (h) | AUClast (h*ng/mL) | AUClast_D (h*kg*ng/mL/mg) | AUCINF_obs (h*ng/mL) | AUCINF_D_obs (h*kg*ng/mL/mg) | Vss_obs (mL/kg) | Cl_obs (mL/h/kg) 

(ex mouse 62) 0.993 ; 0.132 ; 1070 ; 1070 ; 1740 ; 1.00 ; 347;  347 ; 349 ; 349 ; 523 ; 2870 
(mean 1mg/kg) 0.985 ; 0.135 ; 1120 ; 1120 ; 1940 ; 1.00 ; 364 ; 364 ; 366 ; 366 ; 497 ; 2780  
(mean 10mg/kg) 0.668 ; 0.635 ; 8510 ; 851 ; 14900 ; 6.67 ; 3130 ; 313 ; 3130 ; 313 ; 934 ; 3230 

NCA PO:
Rsq_adjusted | HL_Lambda_z (h) | Cmax | Tmax(h) | | Cmax_D (kg*ng/mL/mg) | AUClast (h*ng/mL) | AUClast_D (h*kg*ng/mL/mg) |  AUCINF_obs (h*ng/mL) | AUCINF_D_obs (h*kg*ng/mL/mg) | Cl_F_pred
 
(ex mouse 67)  - ; - ; 1.00; 8.46 ; 0.846 ; 1.48 ; 1.72 ; 17.2  ; - ; - ; - 
(mean 10mg/kg) 0.786 ; 1.15 ; 6.65 ; 0.750 ; 0.665 ; 10.4 ; 1.04 ; 10.3 ; 1.03 

Thank you so much for your valuable feedback!

**Reply by billdenney** [reply]
I'm confused about your results.  When looking at the results for mouse 62, the Cmax_D value from Phoenix is 735, but the observed Cmax at time 0.083 is 1070 and the dose is 1 mg/kg.  And, Clast is 14.6 from Phoenix but the observed Clast is 8.16.

Given the mismatch between the data and the simple to extract results, I'm not sure how I can help.

**Reply by caribap** [reply]
you are completely right! I copied the wrong results. I am sorry for my mistake. My question now doesn't lie so much on the results per si, as with this structure of data I am not even able to get results with PKNCA. Plus, I wanted to understand if I should continue to use sparse as I don't think this structure requires.

Thank you so much for your guidance!

**Reply by caribap** [reply]
My strategy now lies around the fact that I am extracting both concentration and dose info from the same set of data and the time associated to both is extracted from the same column. In your examples you use d_conc whith a different set of time date from d_dose, and I was wondering it that could be the problem to my lack of results, as if I am reading the data wrongly.
Steps to resolution:
- I already split my data by Route, so performing to NCA analysis separately (as I do in phoenix) so I can have one set for extravascular and the other for intravascular - but this doesn't resolved my problems.  
- I already follow your advice to set sparse to TRUE although I don't use this on phoenix - it didn't work.
-  create 2 data frames from my original set of data, where in my dose data I only state:
Mouse  | Time  |  Dose
62            0           1
63            0           1
64            0           1
65            0           10
66            0           10
67            0           10
(...)

Thank you for your patience.

**Reply by caribap** [reply]
the code I wrote: # define tools version + version check
Rversion     = "4.3.2"

Check_Rv <- R.Version()
Check_Rv <- paste0(Check_Rv$major,".",Check_Rv$minor)

if (Rversion == Check_Rv) {
  print("Successful check, all good to proceed")
  
} else {
  stop("Execution stopped. Check Rversion")
}

------ Environment preparation -----------------------------------------------------------------------

--Set the working directory
setwd("...")

--Define the R script file path
rscript_path <- file.path(getwd(), "PK_analysis.R")

--Specify the output file path
output_path <- file.path(getwd(), "PK_analysis.png")

--Check working directory
getwd()

--Load dataset
pk_data <- read_excel("test1.xlsx", skip = 0)

--Visualise the first five rows of my data
head(pk_data)



---------- Data Preparation ---------------------------------------------------------------------------

--Assuming 'IV' for intravascular and 'PO' for extravascular Routes
iv_data <- pk_data %>% filter(Route == "IV")
oral_data <- pk_data %>% filter(Route == "PO")

--Convert values to numeric for both datasets
iv_data$Concentration <- as.numeric(iv_data$Concentration)
iv_data$Dose <- as.numeric(iv_data$Dose)
iv_data$Time <- as.numeric(iv_data$Time)

oral_data$Concentration <- as.numeric(oral_data$Concentration)
oral_data$Dose <- as.numeric(oral_data$Dose)
oral_data$Time <- as.numeric(oral_data$Time)

-- Replace NA values resulting from BLQ with zero
iv_data$Concentration[is.na(iv_data$Concentration)] <- 0
oral_data$Concentration[is.na(oral_data$Concentration)] <- 0


--Filter dose data to include only rows where DoseTime is equal to 0
d_dose <- iv_data[iv_data$DoseTime == 0, c("Subject", "Dose", "DoseTime")]

--Remove rows with missing values
d_dose <- na.omit(d_dose)

--Convert Dose and DoseTime to numeric
d_dose$Dose <- as.numeric(d_dose$Dose)
d_dose$DoseTime <- as.numeric(d_dose$DoseTime)

------------------------------------------------------------------------

o_conc <- PKNCAconc(data=iv_data, formula=Concentration~Time|Subject)

--Put your filtered dose data into a PKNCAdose object
o_dose <- PKNCAdose(data=d_dose, formula=Dose~DoseTime|Subject)

--Combine the two (and automatically determine the intervals of interest)
o_data <- PKNCAdata(o_conc, o_dose)

--Compute the NCA parameters
o_results <- pk.nca(o_data)

--Summarize the results
summary(o_results)

versus the results:
start end N auclast       cmax                    tmax     half.life aucinf.obs
     0  24 6      NC          .                       .             .          .
     0 Inf 6       . 3080 [157] 0.0830 [0.0830, 0.0830] 0.429 [0.344]         NC

Caption: auclast, cmax, aucinf.obs: geometric mean and geometric coefficient of variation; tmax: median and range; half.life: arithmetic mean and standard deviation

Warning messages:
1: Requesting an AUC range starting (0) before the first measurement (0.083) is not allowed

I don't see how to make this package work.

Thank you.

**Comment by smouksassi** [comment]
https://cran.r-project.org/web/packages/PKNCA/vignettes/v04-sparse.html
are you specifying the sparse NCA ?

**Reply by billdenney** [reply]
That's a good question, if your data are sparse, you must indicate that to PKNCA.

**Reply by caribap** [reply]
I am not, I thought this was not an example when I had to use sparse, maybe I understood it wrong. I have in fact 3 mice that are administered only one dose and during a 24h interval suffer 5 plasma sample collection each to measure the concentration over this time slot. I tried to change it to:

iv_conc <- PKNCAconc(iv_data, Concentration~Time|Subject+Dose, sparse=TRUE)
head(iv_conc)

iv_dose <- PKNCAdose(iv_data, Dose~Time|Subject+Dose, "intravascular", sparse=TRUE)
iv_dose

d_interval <- data.frame(start = 0.083,
                         end = 24,
                         c0 = TRUE,
                         cmax = TRUE,
                         tmax = TRUE,
                         Tlast = TRUE,
                         aucivlast = TRUE,
                         aucinf.obs = TRUE
                         # clr.obs = TRUE,
                         # half.life = TRUE,
                         # vd.obs = TRUE
                         )

iv_data <- PKNCAdata(iv_conc,iv_dose, intervals=d_interval)

o_result <- pk.nca(iv_data)
o_result

But I still get this Error: Error in standardize_column_names(x = data_no_exclude, cols = cols, group_cols = group_cols,  : 
  group_cols must not overlap with other column names.  Change the name of the following groups: Dose

**Reply by billdenney** [reply]
Sorry for the slow reply on this.  You cannot use a column name for both a grouping variable and one of the variables like time, concentration, or dose.  The reason for this is that under the surface, PKNCA uses the `dplyr::nest()` function to create groups.  Columns that are used for groups are left outside the data available for calculation.

**Reply by billdenney** [reply]
If you can think of a clearer way to express that error message, please let me know.
